homeassistant/components/bmw_connected_drive/__init__.py

```diff
@@ -45,48 +52,155 @@
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in BMW_COMPONENTS
+                if component != "notify"
+            ]
+        )
+    )
+
+    # Only remove services if it is the last account and not read only
+    if len(hass.data[DOMAIN]) == 1 and not hass.data[DOMAIN][entry.entry_id].read_only:
+        services = list(_SERVICE_MAP) + [SERVICE_UPDATE_STATE]
+        unload_services = all(
+            await asyncio.gather(
+                *[
+                    hass.async_add_executor_job(hass.services.remove, DOMAIN, service)
+                    for service in services
+                ]
+            )
+        )
+    else:
+        unload_services = True
+
+    # Remove notify services
+    unload_notify = unload_services = all(
+        await asyncio.gather(
+            *[
+                hass.async_add_executor_job(
+                    hass.services.remove,
+                    NOTIFY_DOMAIN,
+                    slugify(f"{DOMAIN}_{vehicle.name}"),
+                )
+                for vehicle in hass.data[DOMAIN][entry.entry_id].account.vehicles
+            ]
+        )
+    )
+
+    if all([unload_ok, unload_services, unload_notify]):
+        hass.data[DOMAIN].pop(entry.entry_id)
+    return unload_ok
+
+
+async def update_listener(hass, config_entry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+def setup_account(entry: ConfigEntry, hass, name: str) -> "BMWConnectedDriveAccount":
     """Set up a new BMWConnectedDriveAccount based on the config."""
-    username = account_config[CONF_USERNAME]
-    password = account_config[CONF_PASSWORD]
-    region = account_config[CONF_REGION]
-    read_only = account_config[CONF_READ_ONLY]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    region = entry.data[CONF_REGION]
+    read_only = entry.options[CONF_READ_ONLY]
+    use_location = entry.options[CONF_USE_LOCATION]
 
     _LOGGER.debug("Adding new account %s", name)
-    cd_account = BMWConnectedDriveAccount(username, password, region, name, read_only)
 
-    def execute_service(call):
-        """Execute a service for a vehicle.
+    pos = (
+        (hass.config.latitude, hass.config.longitude) if use_location else (None, None)
+    )
+    cd_account = BMWConnectedDriveAccount(
+        username, password, region, name, read_only, pos[0], pos[1]
+    )
 
-        This must be a member function as we need access to the cd_account
-        object here.
-        """
+    def execute_service(call):
+        """Execute a service for a vehicle."""
         vin = call.data[ATTR_VIN]
-        vehicle = cd_account.account.get_vehicle(vin)
+        vehicle = None
+        # Double check for read_only accounts as another account could create the services
+        for account in [
```

Comment:
```
for account in filter(lambda account: not account.read_only, hass.data[DOMAIN]):

Reply:
Is this faster/better than the list comprehension? Haven't seen this before so to me it looks more complicated/not as easy to read.

Reply:
it is faster because you only iterate once